File: wtwitchParse/wtwitchParse.nim
import std/json
import std/strutils

# Declarations
const
    ERR = "Error fetching subs."
    NONE = "None live."
    BAD_ICON = "bell-slash"
    GOOD_ICON = "bell"

type
    Streamer = object
        user_name: string

# Read data and create output
var output: string
try:
    let data = readFile("~/.cache/wtwitch/subscription-cache.json")
    let jsonNode = parseJson(data)["data"]

    doAssert len(jsonNode) != 0
    var live: seq[string]
    for streamer in jsonNode:
        let curr_streamer = streamer.to(Streamer)
        live.add(curr_streamer.user_name)
    output = join(live, " ")
except AssertionDefect:
    output = ""
except:
    output = ERR

# JSON output creation
var json_output: JsonNode
if output != "": # ERR or normal
    let state = if output == ERR: "Critical" else: "Good"
    let icon = if output == ERR: BAD_ICON else: GOOD_ICON
    json_output = %*{"icon": icon, "text": output, "state": state}
else:
    json_output = %*{"text": NONE, "icon": BAD_ICON}
echo json_output

File: wtwitchParse/wtwitchParse.rs
use serde_json::json;
use serde_json::Value;
use std::fs;

fn main() {
    const ERR: &str = "Error fetching subs.";

    let data = fs::read_to_string("/home/jean-leon/.cache/wtwitch/subscription-cache.json");
    let res: String = match data {
        Ok(content) => content,
        Err(_error) => ERR.to_string(),
    };

    let mut output = String::from("");
    if res != ERR {
        let v: Value = serde_json::from_str(&res).unwrap_or(json!("parsing error"));
        for i in 1..v["data"].as_array().unwrap().len() + 1 {
            let streamer = v["data"][i - 1]["user_name"].as_str().unwrap_or("");
            output.push_str(streamer);
            output.push_str(" ");
        }
        output = output.trim_end().to_string();
    } else {
        output = res;
    }

    if output != "" {
        let state = if output == ERR { "Critical" } else { "Good" };
        let icon = if output == ERR { "bell-slash" } else { "bell" };
        print!(
            "{}",
            json!(
                {
                    "icon" : icon,
                    "text" : format!("{}", output),
            "state" : state,
                })
            .to_string()
        );
    } else {
        print!("{}", json!({"text":"None live.", "icon":"bell-slash"}));
    }
}
